### ISSUE TITLE
Clarifications on sub-group support

### ISSUE THREAD

**Issue by frasercrmck** [KhronosGroup/SYCL-Docs#463]
This is generally seeking clarifications about the requirements made on a device's support for sub-groups.

In [4.6.4.2](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_device_information_descriptors), `info::device::sub_group_sizes` says that it:

``` quote
Returns a std::vector of size_t containing the set of sub-group sizes supported by the device.
````

I think the spec could use clarification about what the minimal set of sub-group sizes looks like. Can a device return an empty vector? What would that mean?

The SYCL-CTS assumes the vector is non-empty, as discussed in  [SYCL-CTS/pull/797](https://github.com/KhronosGroup/SYCL-CTS/pull/797).

Furthermore, the use of "**the** set" indicates to me that this returns _all_ of the possible sub-group sizes the compiler ever generates. Is this the case? Is it incorrect if an implementation's compiler generates a kernel with a sub-group size not in turn advertised by `info::device::sub_group_sizes`? Again, the SYCL-CTS does currently mandate this.

I think the current spec wording infers that there's at least one sub-group size and that the reported set of sub-group must be complete and total, but I think this could be better clarified.

As a bit of background/context for my next questions/suggestions, our implementation (the oneAPI construction kit) is not a hardware vendor, per se: we're a compiler toolkit that hopes to support as many devices as possible.

One of our methods of supporting sub-groups for devices was OpenCL's concept of "degenerate sub-groups" - that there's one sub-group covering the entire work-group. This doesn't seem to have a home in SYCL, as `info::device::sub_group_sizes` would in effect return all possible work-group dimensions. Then, in turn, a user would feasibly "require" one of those sub-group sizes via `reqd_sub_group_size` which doesn't make any sense, either: a kernel's sub-group size becomes a property of the dynamic execution range, not the device.

Should SYCL have better support for "degenerate" sub-groups?

Another of our methods of supporting sub-groups is via whole-function vectorization. Groups of work-items run together via vectorization form a sort of sub-group. So in that sense we are also able to support any number of vectorized sub-group sizes - we're not tied to any hardware and the compiler is able to choose the best one. This again is awkward to report via `info::device::sub_group_sizes` as there's a lot of sizes, but it's not as bad as the degenerate case.

Vectorized sub-groups make sense to me via `reqd_sub_group_size` as we can force vectorization to that width, but what if the user doesn't specify one? Must the compiler then vectorize to only one of the sizes reported by `info::device::sub_group_sizes`? Is it **the** set of sizes, after all. That feels a bit limiting, but I understand that the user wants better control over sub-groups, as there are problems with how open the OpenCL spec left it.

Overall it would be nice if SYCL had better support for sub-groups that aren't tied to a specific hardware vendor, as they appear to me to currently be. I'd appreciate anyone's thoughts on this matter!

**Comment by Pennycook** [comment]
> I think the spec could use clarification about what the minimal set of sub-group sizes looks like. Can a device return an empty vector? What would that mean?

I agree we should clarify this. I think the vector should contain at least one sub-group size.

> I think the current spec wording infers that there's at least one sub-group size and that the reported set of sub-group must be complete and total, but I think this could be better clarified.

In my opinion, this is the correct behavior. Most devices I'm familiar with prefer sub-group sizes that are powers of 2; limiting even to something like the first 10 powers of 2 would be sufficient to support very large sub-group sizes.

> Should SYCL have better support for "degenerate" sub-groups?
 
I think "degenerate" sub-groups in OpenCL are weird, and lead to inconsistency in the way that people think about sub-groups.  There are functions that OpenCL (and SYCL) only make available for sub-groups (like shuffles, non-uniform groups, etc) -- presumably because implementing those functions for work-groups would be too difficult/expensive. But with "degenerate" sub-groups, these functions are (effectively) available for work-groups as well.

SYCL supports a different "degenerate" case, with sub-groups of size 1.  These are trivial to implement (e.g., barriers are no-ops, reductions return the input value unmodified).  Any device can support a sub-group size of 1 for all kernels, and this is more predictable than a sub-group size which changes frequently (to match the work-group size).
 
> Vectorized sub-groups make sense to me via `reqd_sub_group_size` as we can force vectorization to that width, but what if the user doesn't specify one? Must the compiler then vectorize to only one of the sizes reported by `info::device::sub_group_sizes`? Is it **the** set of sizes, after all. That feels a bit limiting, but I understand that the user wants better control over sub-groups, as there are problems with how open the OpenCL spec left it.

We have a [proposal](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc) to clarify that the compiler must vectorize to only one of the sizes reported by `sub_group_sizes`, and to additionally require that the default behavior is that all kernels use the _same_ sub-group size.  We have had feedback from multiple developers that dealing with multiple sub-group sizes is too confusing.

**Comment by frasercrmck** [comment]
> > I think the spec could use clarification about what the minimal set of sub-group sizes looks like. Can a device return an empty vector? What would that mean?
> 
> I agree we should clarify this. I think the vector should contain at least one sub-group size.

Makes sense to me.


> > Should SYCL have better support for "degenerate" sub-groups?
> 
> I think "degenerate" sub-groups in OpenCL are weird, and lead to inconsistency in the way that people think about sub-groups. There are functions that OpenCL (and SYCL) only make available for sub-groups (like shuffles, non-uniform groups, etc) -- presumably because implementing those functions for work-groups would be too difficult/expensive. But with "degenerate" sub-groups, these functions are (effectively) available for work-groups as well.

I agree - I don't like degenerate sub-groups (as an implementer, not even as a user) and they really don't seem to be viable in SYCL for the reasons you suggest, and more.

> SYCL supports a different "degenerate" case, with sub-groups of size 1. These are trivial to implement (e.g., barriers are no-ops, reductions return the input value unmodified). Any device can support a sub-group size of 1 for all kernels, and this is more predictable than a sub-group size which changes frequently (to match the work-group size).

That makes sense to me. I think the spec would benefit from clarifying this, in combination with your first suggestion. Not just because I imagine people will compare and contrast the SYCL and OpenCL specifications, and leaving the term "degenerate" ambiguous could cause confusion.

> > Vectorized sub-groups make sense to me via `reqd_sub_group_size` as we can force vectorization to that width, but what if the user doesn't specify one? Must the compiler then vectorize to only one of the sizes reported by `info::device::sub_group_sizes`? Is it **the** set of sizes, after all. That feels a bit limiting, but I understand that the user wants better control over sub-groups, as there are problems with how open the OpenCL spec left it.
> 
> We have a [proposal](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc) to clarify that the compiler must vectorize to only one of the sizes reported by `sub_group_sizes`, and to additionally require that the default behavior is that all kernels use the _same_ sub-group size. We have had feedback from multiple developers that dealing with multiple sub-group sizes is too confusing.

I think restricting the list of sub-group sizes to one of the device's reported sizes makes sense. We have already started doing that.

I'm a bit worried about "all kernels use the same preferred sub-group size". As an implementation (or rather, toolkit) which relies on vectorization as a means of providing sub-group sizes, I would never be able to guarantee that we'll be able to vectorize an arbitrary kernel. If we can't vectorize one in a thousand, does that mean we can't vectorize any?

Would it be possible to provide a fall-back to the "degenerate" case? Either they're all the preferred size, or size 1?

**Comment by Pennycook** [comment]
> I'm a bit worried about "all kernels use the same preferred sub-group size". As an implementation (or rather, toolkit) which relies on vectorization as a means of providing sub-group sizes, I would never be able to guarantee that we'll be able to vectorize an arbitrary kernel. If we can't vectorize one in a thousand, does that mean we can't vectorize any?

Whether a SYCL implementation vectorizes a kernel (in the sense that it generates SIMD instructions) is not observable to a SYCL developer, so I don't think you have to worry about this guarantee.  If the sub-group member functions and group algorithms are implemented in a way that meets the specification, then the kernel will appear to be using a specific sub-group size, even if it's not using vector instructions.  This gives an implementation the freedom to fall back to _S_ separate scalar instructions in place of a single vector instruction (if there's just one operation that can't be vectorized), or to completely serialize execution of work-items between sub-group barriers and algorithms.


> Would it be possible to provide a fall-back to the "degenerate" case? Either they're all the preferred size, or size 1?

I think this would make things too complicated for the developer.  The main motivations for requiring a single default sub-group size are:

- Most developers won't ever need to specify a sub-group size on kernel definitions or launches.
- Developers won't need to query which sub-group size a kernel actually used, simply to check if an implementation generated something different to what they were expecting.
- The default sub-group size can be used to define data structures (e.g., in Array-of-Structure-of-Arrays format), to improve vectorization and/or cache behavior for sub-groups.

In my opinion, allowing a fallback case where the primary sub-group size isn't used would defeat the point of introducing a primary sub-group size.  It would be much nicer for developers if SYCL implementations just made this work.
